benchmarks/dataset_serialize_benchmark.py

```diff
@@ -0,0 +1,232 @@
+import itertools
+import logging
+import os
+import shutil
+import time
+import uuid
+
+import conbench.runner
+import pyarrow
+import pyarrow.dataset as ds
+
+from benchmarks import _benchmark
+
+log = logging.getLogger(__name__)
+
+
+@conbench.runner.register_benchmark
+class DatasetSerializeBenchmark(_benchmark.Benchmark):
+    """
+    This benchmark is supposed to measure the time it takes to write data from
+    memory (from a pyarrow Table) to a tmpfs file system, given a specific
+    serialization format (parquet, arrow, ...).
+
+    To make this benchmark agnostic to disk read performance on the input side
+    of things, the data is read fully into memory before starting (and timing)
+    the benchmark function. That is (believed to be) achieved with:
+
+        data = source_dataset.to_table(filter=somefilter, ...)
+
+    After data of interest has been read into memory, the following call is
+    used for both serialization and writing-to-filesystem in one go:
+
+        pyarrow.dataset.write_dataset(format=someformat, ...)
+
+    That operation is timed (and the duration is the major output of this
+    benchmark).
+
+    The data is written to `/dev/shm` (available on all Linux systems). This is
+    a file system backed by RAM (tmpfs). The assumption is that writing to
+    tmpfs is fast (so fast that benchmark duration is significantly affected by
+    the serialization itself), and stable (so that its performance is ~constant
+    across runs on the same machine).
+
+    This benchmark does not resolve how much time goes into the CPU work for
+    serialization vs. the system calls for writing to tmpfs (that would be a
+    different question to answer, an interesting one, that is maybe more of a
+    task for profiling).
+
+    There are two dimensions that are varied:
+
+        - serialization format
+        - amount of the data being written, as set by a filter on the input
+    """
+
+    name = "dataset-serialize"
+
+    arguments = ["source"]
+
+    sources = [
+        "nyctaxi_multi_parquet_s3",
+        # "nyctaxi_multi_ipc_s3",
+        # "chi_traffic_2020_Q1",
+    ]
+
+    sources_test = [
+        "nyctaxi_multi_parquet_s3_sample",
+        "nyctaxi_multi_ipc_s3_sample",
+        "chi_traffic_sample",
+    ]
+
+    _params = {
+        "selectivity": ("1pc", "10pc"),
+        "format": (
+            "parquet",
+            "arrow",
+            "ipc",
+            "feather",
```

Comment:
Great input.

> arrow == feather, at least for feather v2, 

Interesting! Hmm, the timings that I saw suggest that this equality isn't quite right. It appears like `feather` and `ipc` are highly comparable, e,g.:

```
dataset: nyctaxi_multi_parquet_s3, format: arrow, selectivity: 100pc
    --> min: 1.38 s     mean±SE: (1.48 ± 0.05) s
dataset: nyctaxi_multi_parquet_s3, format: ipc, selectivity: 100pc
    --> min: 1.39 s     mean±SE: (1.55 ± 0.09) s
dataset: nyctaxi_multi_parquet_s3, format: feather, selectivity: 100pc
    --> min: 1.44 s     mean±SE: (1.55 ± 0.06) s
```

```
dataset: nyctaxi_multi_ipc_s3, format: arrow, selectivity: 100pc
    --> min: 1.64 s     mean±SE: (1.77 ± 0.08) s
dataset: nyctaxi_multi_ipc_s3, format: ipc, selectivity: 100pc
    --> min: 1.75 s     mean±SE: (2.4 ± 0.2) s
dataset: nyctaxi_multi_ipc_s3, format: feather, selectivity: 100pc
    --> min: 2.34 s     mean±SE: (2.7 ± 0.1) s
```

https://arrow.apache.org/docs/python/feather.html

> Version 2 (V2), the default version, which is **exactly represented as the Arrow IPC file** format on disk. V2 files support storing all Arrow data types as well as compression with LZ4 or ZSTD. V2 was first made available in Apache Arrow 0.17.0.

So, I will remove either `feather` or `ipc`.

The [docs for write_dataset()](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html) suggest that all three are the same/comparable:

> The format in which to write the dataset. Currently supported: “parquet”, “ipc”/”arrow”/”feather”, and “csv”. 

However, there probably is a difference (as indicated by measurement results). I think it might have to do with compression. For `feather`, this [is documented](https://arrow.apache.org/docs/python/feather.html):

> LZ4 is used by default if it is available (which it should be if you obtained pyarrow through a normal package manager):

Maybe maybe `feather` is `lz4(arrow)`.

I think in this case it makes sense to leave in both, kind of covering the 'default' settings.